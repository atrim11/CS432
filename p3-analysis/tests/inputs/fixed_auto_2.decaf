// Function definitions for testing
def int subtract(int x, int y) {
    return x - y;
}

def void printNumber(int num) {
    print_int(num);  // Error: should use print_int instead of print_str
}

def int concatStrings(int a, int b) {
    return a + b;  // Error: In Decaf, strings cannot be concatenated with '+'
}

int arr[12];
def int main() {
    int result;
    int a;
    int b;
    int test;
    int c;
    result = subtract(10, 5);
    printNumber(result);  // Error: wrong type to print

    // Parameter mismatch
    result = subtract(10, 5);  // Error: subtract expects 2 parameters, provided 1

    // Type mismatch
    result = subtract(1, 5);  // Error: first parameter should be an int

    // Recursive issues
    result = subtract(subtract(20, 10), 31);  // Error: passing string instead of int

    // Incorrect return type
    
    // Redeclaration of variables
    a = 5;
    a = 10;  // Error: redeclaration of variable 'a' in the same scope

    // Undefined variables
    b = a + c;  // Error: 'c' is not declared
    
    // Array out of bounds
    arr[3] = 100;  // Error: out of bounds

    // Function with wrong return type
    
    // Control structure issues
    if (true) {  // Error: non-boolean expression
        print_str("This is wrong");
    }

    while (1 + 5 >= 3) {  // Error: non-boolean used in while loop
        result = result - 1;
    }

    // Incorrect use of void function

    // More mismatches and errors
    test = subtract(subtract(52, 5), 10);  // Error: multiple type mismatches
    return test;
}
