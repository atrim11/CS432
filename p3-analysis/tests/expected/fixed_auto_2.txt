Program [line 1]
SYM TABLE:
 print_int : (int) -> void
 print_bool : (bool) -> void
 print_str : (str) -> void
 subtract : (int, int) -> int
 printNumber : (int) -> void
 concatStrings : (int, int) -> int
 main : () -> int
 arr : int [12]

  FuncDecl name="subtract" return_type=int parameters={x:int,y:int} [line 2]
  SYM TABLE:
   x : int
   y : int

    Block [line 2]
    SYM TABLE:

  FuncDecl name="printNumber" return_type=void parameters={num:int} [line 6]
  SYM TABLE:
   num : int

    Block [line 6]
    SYM TABLE:

  FuncDecl name="concatStrings" return_type=int parameters={a:int,b:int} [line 10]
  SYM TABLE:
   a : int
   b : int

    Block [line 10]
    SYM TABLE:

  FuncDecl name="main" return_type=int parameters={} [line 15]
  SYM TABLE:

    Block [line 15]
    SYM TABLE:
     result : int
     a : int
     b : int
     test : int
     c : int

        Block [line 48]
        SYM TABLE:

        Block [line 52]
        SYM TABLE:

